### PR TITLE
Support for online FilesCollection

### PR DESCRIPTION
**Pull request by Flowtter**
+ add support for aws S3 bucket to store files
  - upload file on the bucket
  - retrieve file on the bucket
  - delete file
+ add support for a Content delivery network
  - retrieve file of the bucket using a CDN

### CODE REVIEW

**Comment by ramnes** [comment]
Just adding that the overall point of using S3 to store files is to make lemverse stateless so that it's easier to scale horizontally. :)

File: core/lib/file-storage.js [hunk 1]
```diff
@@ -0,0 +1,201 @@
+// Taken from https://github.com/VeliovGroup/Meteor-Files/blob/master/docs/aws-s3-integration.md
+import { Meteor } from 'meteor/meteor';
+import { _ } from 'meteor/underscore';
+import { Random } from 'meteor/random';
+import { FilesCollection } from 'meteor/ostrio:files';
+import stream from 'stream';
+
+import S3 from 'aws-sdk/clients/s3'; /* http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html */
```
Comment (ramnes) [inline core/lib/file-storage.js:8]: Looks like you forgot to add aws-sdk to the deps @Flowtter :)

Comment (xsyann) [reply core/lib/file-storage.js:8]: Fixed :)

File: core/lib/file-storage.js [hunk 2]
```diff
@@ -0,0 +1,201 @@
+// Taken from https://github.com/VeliovGroup/Meteor-Files/blob/master/docs/aws-s3-integration.md
+import { Meteor } from 'meteor/meteor';
+import { _ } from 'meteor/underscore';
+import { Random } from 'meteor/random';
+import { FilesCollection } from 'meteor/ostrio:files';
```
Comment (ramnes) [inline core/lib/file-storage.js:5]: Don't we need to add a dep for this one as well?

Comment (Flowtter) [reply core/lib/file-storage.js:5]: The package was already in: https://github.com/l3mpire/lemverse/blob/19a80c88c939892a865bc2f82d0182ec944c8e66/app/.meteor/packages#L35

Yet some files have been removed during my rebase. I'll fix it ASAP.

File: core/lib/file-storage.js [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+/* For better i/o performance */
+import fs from 'fs';
+
+
+import { fileOnBeforeUpload } from './misc';
+
+function fileSystemAdapter() {
+  const s3Conf = Meteor.settings.s3;
+
+  if (!s3Conf) {
```
Comment (Donorhan) [inline core/lib/file-storage.js:19]: A non-blocking detail, wouldn't it be better to put the empty S3 config in all settings files and check instead that the "s3?.key" attribute exists?

**Comment by Donorhan** [comment]
Thanks for your contribution, it will be useful to many people I think 🙂

(I left a comment but everything looks great to me! If you don't think it's necessary, I'll merge the PR)

**Comment by Flowtter** [comment]
I'll have a look soon!

File: core/lib/file-storage.js (hunk 3 above)
Comment (Flowtter) [reply core/lib/file-storage.js]: Fixed in https://github.com/l3mpire/lemverse/commit/9299022b70527507c397099c35cf708ad9119a9b

**Comment by Donorhan** [comment]
Thanks! 🙏🏻

**Review by Donorhan — APPROVED** [review]

**Pull request merged**